Adding waveforms + refactoring mem docs (#84)
freechipsproject/www.chisel-lang.org: docs/src/main/tut/chisel3/memories.md

### `SyncReadMem`: sequential/synchronous-read, sequential/synchronous-write

Chisel has a construct called `SyncReadMem` for sequential/synchronous-read, sequential/synchronous-write memories. These `SyncReadMem`s will likely be synthesized to technology SRAMs (as opposed to register banks).

If the same memory address is both written and sequentially read on the same clock edge, or if a sequential read enable is cleared, then the read data is undefined.

Values on the read data port are not guaranteed to be held until the next read cycle. If that is the desired behavior, external logic to hold the last read value must be written.

#### Read port/write port
Ports into `SyncReadMem`s are created by applying a `UInt` index.  A 1024-entry register file with one write port and one read port might be expressed as follows:

> ```scala

Reviewer: ```suggestion
```scala mdoc:silent
import chisel3._
```

where `amp` is used to scale the fixpoint values stored in the ROM.

## Memories

Memories are given special treatment in Chisel since hardware implementations of memory vary greatly. For example, FPGA memories are instantiated quite differently from ASIC memories. Chisel defines a memory abstraction that can map to either simple Verilog behavioural descriptions or to instances of memory modules that are available from external memory generators provided by foundry or IP vendors.

### `SyncReadMem`: sequential/synchronous-read, sequential/synchronous-write

Chisel has a construct called `SyncReadMem` for sequential/synchronous-read, sequential/synchronous-write memories. These `SyncReadMem`s will likely be synthesized to technology SRAMs (as opposed to register banks).

If the same memory address is both written and sequentially read on the same clock edge, or if a sequential read enable is cleared, then the read data is undefined.

> Values on the read data port are not guaranteed to be held until the next read cycle. If that is the desired behavior, external logic to hold the last read value must be written.

Reviewer 2: ```suggestion
Values on the read data port are not guaranteed to be held until the next read cycle. If that is the desired behavior, external logic to hold the last read value must be added.
```

"written" is a little confusing in the context of SRAMs with all the read/writing going on. Minor suggestion though don't need to block on this

Author: oh yeah, that is confusing

#### Read port/write port

> Ports into `SyncReadMem`s are created by applying a `UInt` index.  A 1024-entry register file with one write port and one read port might be expressed as follows:

Reviewer 2: ```suggestion
Ports into `SyncReadMem`s are created by applying a `UInt` index.  A 1024-entry SRAM with one write port and one read port might be expressed as follows:
```
Or do you mean SRAM/register file? I'm really not sure since the contrast is with `Mem` which we. say usually would be a register file

Author: yes sorry, it should be SRAM here.
